packages/drawer/src/Drawer.tsx

```diff
@@ -55,30 +62,48 @@ const Drawer = ({
   hideCancel = false,
 }: DrawerProps) => {
   const ref = useRef<HTMLDivElement>(null);
+  const [isFocusWithin, setIsFocusWithin] = useState(false);
+  const { focusWithinProps } = useFocusWithin({
+    onFocusWithinChange: (isFocusWithin) => setIsFocusWithin(isFocusWithin),
+  });
 
   usePreventScroll();
 
   useEffect(() => {
     const handleEscape = (event: KeyboardEvent) => {
       event.stopImmediatePropagation();
       const latest = [...document.querySelectorAll('[data-drawer]')].pop();
-      if (event.key === 'Escape' && latest === ref.current) {
+      if (event.key === 'Escape' && latest === ref.current && isFocusWithin) {
         close();
       }
     };
 
+    const addOverlayAndEventHandler = () => {
+      document.body.classList.add('has-overlay');
+      document.addEventListener('keydown', handleEscape);
+    };
+
+    const removeOverlayAndEventHandler = () => {
+      document.body.classList.remove('has-overlay');
+      document.removeEventListener('keydown', handleEscape);
+    };
+
     const close = () => {
       onCancel?.();
     };
 
-    document.body.classList.add('has-overlay');
-    document.addEventListener('keydown', handleEscape);
+    if (isFocusWithin) {
+      addOverlayAndEventHandler();
+    }
+
+    if (!isFocusWithin) {
+      removeOverlayAndEventHandler();
+    }
 
     return () => {
-      document.body.classList.remove('has-overlay');
-      document.removeEventListener('keydown', handleEscape);
+      removeOverlayAndEventHandler();
     };
-  }, [onCancel, testId]);
+  }, [onCancel, testId, isFocusWithin]);
```

Comment:
this is what allows for smoother drawer + modal focus behavior. adding/removing event listeners based on if the element is focused within